### DISCUSSION TITLE
File and media libraries deduplication

### DISCUSSION THREAD

**Discussion by antonioribeiro** [area17/twill#880] [Ideas]
A nice addition would be not to let people to upload duplicate image files to the libraries. Just compute and store a hash of the file during upload, and search for the hash before adding anything to it.

Maybe also add a configuration key to that too, because sometimes people might want to have duplicates instead.

**Comment by kirkbushell** [comment]
I get what you're doing here, but my experience has shown that this is not always a great idea. For example, if you have a particular image that is used across numerous pieces of content, and you delete that image - all content would be affected, which may not be ideal. I learnt this when we used a similar sort of filesystem mechanic on our SaaS platform - users would delete a file, and not realise they're affecting content elsewhere.

**Comment by antonioribeiro** [comment]
I just sent a [PR](https://github.com/area17/twill/pull/899) to do it. 

Feature is, obviously, disabled by default.

If user's behaviour is a concern, I believe it's not really hard to prevent files from being deleted if `media_id` count on `mediables` is `> 0`. People will have to delete all related content before deleting the media.

**Reply by ifox** [reply]
That's actually already the case, media and files can't be deleted if they are related to anything.

https://github.com/area17/twill/blob/a5b3ffc20ab7826bff6825f28982444e7272537d/src/Models/Media.php#L60
